From abd04121af7cdb23c2054ece751c474a0c1ec1e9 Mon Sep 17 00:00:00 2001
From: Sachin Prabhu <sprabhu@redhat.com>
Date: Mon, 28 Jul 2014 12:16:47 -0400
Subject: [fs] cifs: fix error handling cifs_user_readv

Message-id: <1406549822-10376-55-git-send-email-sprabhu@redhat.com>
Patchwork-id: 86781
O-Subject: [RHEL7.1 PATCH 54/69] [bz1082049]cifs: fix error handling cifs_user_readv
Bugzilla: 1082049
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Jeff Layton <jlayton@redhat.com>

Coverity says:

*** CID 1202537:  Dereference after null check  (FORWARD_NULL)
/fs/cifs/file.c: 2873 in cifs_user_readv()
2867       cur_len = min_t(const size_t, len - total_read, cifs_sb->rsize);
2868       npages = DIV_ROUND_UP(cur_len, PAGE_SIZE);
2869
2870       /* allocate a readdata struct */
2871       rdata = cifs_readdata_alloc(npages,
2872              cifs_uncached_readv_complete);
>>>     CID 1202537:  Dereference after null check  (FORWARD_NULL)
>>>     Comparing "rdata" to null implies that "rdata" might be null.
2873       if (!rdata) {
2874        rc = -ENOMEM;
2875        goto error;
2876       }
2877
2878       rc = cifs_read_allocate_pages(rdata, npages);

...when we "goto error", rc will be non-zero, and then we end up trying
to do a kref_put on the rdata (which is NULL). Fix this by replacing
the "goto error" with a "break".

Reported-by: <scan-admin@coverity.com>
Signed-off-by: Jeff Layton <jlayton@redhat.com>
Signed-off-by: Steve French <smfrench@gmail.com>
(cherry picked from commit bae9f746a18ee31bbeeb25ae6615805ed6eca173)
Signed-off-by: Sachin Prabhu <sprabhu@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/cifs/file.c b/fs/cifs/file.c
index 74601b9..483271c 100644
--- a/fs/cifs/file.c
+++ b/fs/cifs/file.c
@@ -2909,7 +2909,7 @@ cifs_iovec_read(struct file *file, const struct iovec *iov,
          cifs_uncached_readv_complete);
   if (!rdata) {
    rc = -ENOMEM;
-   goto error;
+   break;
   }
 
   rc = cifs_read_allocate_pages(rdata, npages);
-- 
1.7.1